app/models/upload.rb

```diff
@@ -70,28 +76,76 @@ def count_contents(upload, meta)
     else
       upload_path = submission_path.to_s
     end
+    output, err, status = "", "", ""
     begin
       if meta[:mimetype] == "application/zip" || upload_path.ends_with?(".zip")
-        return ZipRuby::Archive.open(upload_path).num_files
+        offset = 5 # archive name, column headers, dashes, dashes, and summary lines
+        output, status = Open3.capture2("unzip", "-l", upload_path)
       elsif meta[:mimetype] == "application/x-tar" || upload_path.ends_with?(".tar")
-        return SubTarball.count_files(upload_path)
+        output, err, status = Open3.capture3("tar", "-tvf", upload_path)
+        offset = 0
       elsif meta[:mimetype] == "application/x-compressed-tar" || upload_path.ends_with?(".tgz")
-        return SubTarball.count_files_gz(upload_path)
+        output, err, status = Open3.capture3("tar", "-tzvf", upload_path)
+        offset = 0
       elsif meta[:mimetype] == "application/gzip" || upload_path.ends_with?(".gz")
         if submission_path.to_s.ends_with?(".tar.gz")
-          return SubTarball.count_files_gz(upload_path)
+          output, err, status = Open3.capture3("tar", "-tf", upload_path)
+          offset = 0
         else
           return 1
         end
       else
         return 1
       end
+      raise Exception.new("Could not list the files in #{upload_path}") if !status.success?
+      return output.lines.count - offset
     rescue Exception => e
       puts e.message
       raise Exception.new("Could not read archive to count files")
     end
   end
 
+  def total_size_too_large?(upload, meta)
+    if upload.is_a? ActionDispatch::Http::UploadedFile
+      upload_path = upload.path
+    elsif upload.is_a? String
+      upload_path = upload
+    else
+      upload_path = submission_path.to_s
+    end
+    begin
+      if meta[:mimetype] == "application/zip" || upload_path.ends_with?(".zip")
+        output, err, status = Open3.capture3("unzip", "-l", upload_path)
```

Comment:
I don't know.  I've seen one misleading .zip file whose reported sizes are way too _large_; I don't know if it's possible for them to be misreported as too _small_.